[0.17] Add pegin validation

stevenroose opened the pull request.

instagibbs commented on src/init.cpp:
```diff
@@ -521,8 +521,30 @@ void SetupServerArgs()
+    gArgs.AddArg("-secretprefix", strprintf("The byte prefix, in decimal, of the chain's base58 secret key encoding. (default: %d)", 239), false, OptionsCategory::ELEMENTS);
+    gArgs.AddArg("-extpubkeyprefix", strprintf("The 4-byte prefix, in hex, of the chain's base58 extended public key encoding. (default: %s)", "043587CF"), false, OptionsCategory::ELEMENTS);
+    gArgs.AddArg("-extprvkeyprefix", strprintf("The 4-byte prefix, in hex, of the chain's base58 extended private key encoding. (default: %s)", "04358394"), false, OptionsCategory::ELEMENTS);
+    gArgs.AddArg("-bech32_hrp", strprintf("The human-readable part of the chain's bech32 encoding. (default: %s)", "bc"), false, OptionsCategory::ELEMENTS);
```
because we haven't picked one for elements yet!

In general I expect us to take the network name specifically in the `hrp`, like `liquidv`, but for elements it can be more generic.

`ele`?

stevenroose replied:
I think the base58 defaults are regtest's if I'm not mistaken. So perhaps this should be `"bcrt"`?

stevenroose replied:
Ah no this is not for parent. Yeah this is because we don't have one for elements yet.

luke-jr commented on src/init.cpp:
```diff
@@ -521,8 +521,30 @@ void SetupServerArgs()
+    gArgs.AddArg("-extprvkeyprefix", strprintf("The 4-byte prefix, in hex, of the chain's base58 extended private key encoding. (default: %s)", "04358394"), false, OptionsCategory::ELEMENTS);
+    gArgs.AddArg("-bech32_hrp", strprintf("The human-readable part of the chain's bech32 encoding. (default: %s)", "bc"), false, OptionsCategory::ELEMENTS);
+    gArgs.AddArg("-mainchainrpchost=<host>", "The address which the daemon will try to connect to the trusted bitcoind to validate peg-ins, if enabled. (default: cookie auth)", false, OptionsCategory::ELEMENTS);
+    gArgs.AddArg("-mainchainrpcport=<n>", "The port which the daemon will try to connect to the trusted bitcoind to validate peg-ins, if enabled. (default: cookie auth)", false, OptionsCategory::ELEMENTS);
```
host/port are used even for cookie auth.

luke-jr commented on src/init.cpp:
```diff
@@ -521,8 +521,30 @@ void SetupServerArgs()
+    gArgs.AddArg("-mainchainrpcpassword=<pwd>", "The rpc password which the daemon will use to connect to the trusted bitcoind to validate peg-ins, if enabled. (default: cookie auth)", false, OptionsCategory::ELEMENTS);
+    gArgs.AddArg("-mainchainrpccookiefile=<file>", "The bitcoind cookie auth path which the daemon will use to connect to the trusted bitcoind to validate peg-ins. (default: `<datadir>/regtest/.cookie`)", false, OptionsCategory::ELEMENTS);
+    gArgs.AddArg("-mainchainrpctimeout=<n>", strprintf("Timeout in seconds during mainchain RPC requests, or 0 for no timeout. (default: %d)", DEFAULT_HTTP_CLIENT_TIMEOUT), false, OptionsCategory::ELEMENTS);
+    gArgs.AddArg("-validatepegin", strprintf("Validate peg-in claims. An RPC connection will be attempted to the trusted bitcoind using the `mainchain*` settings below. All functionaries must run this enabled. (default: %u)", DEFAULT_VALIDATE_PEGIN), false, OptionsCategory::ELEMENTS);
```
"below" is wrong

stevenroose replied:
fixed

luke-jr commented on src/chainparams.cpp:
```diff
@@ -476,6 +500,12 @@ class CCustomParams : public CRegTestParams {
+
+        if (!anyonecanspend_aremine) {
+            assert("Anyonecanspendismine was marked as false, but they are in the genesis block"
+                    && initialFreeCoins == 0);
```
```
chainparams.cpp:506:24: error: ‘initialFreeCoins’ was not declared in this scope
```

stevenroose replied:
Yeah, it's a WIP PR, so there's plenty of loose ends still.

luke-jr commented on src/primitives/transaction.h:
```diff
@@ -105,7 +136,69 @@ class CTxIn
+                // index field. The IssuanceMode enum values are chosen to
+                // make this as simple as a bitwise-OR.
+                outpoint.hash = prevout.hash;
+                outpoint.n = prevout.n & COutPoint::OUTPOINT_INDEX_MASK;
```
This is your genesis block problem. A coinbase's outpoint has `n==0xffffffff`, which is being masked off here.

stevenroose replied:
YES! Thanks! I was looking at this line, but somehow assumed the coinbase outpoint was 0..

instagibbs commented on src/primitives/transaction.h:
```diff
@@ -105,7 +136,69 @@ class CTxIn
+        READWRITE(outpoint);
+
+        if (ser_action.ForRead()) {
+            //if (outpoint.n == (uint32_t) -1) {
```
this needs to be active to not have coinbase outpoints be invalid as @luke-jr said

stevenroose commented:
Want to point out explicitly that I removed a check in `AcceptToMempoolWorker` in https://github.com/ElementsProject/elements/pull/458/commits/a9c033195be6cb9b76740dd63e60ae5ef5761ea4.

It seems that that `AcceptToMempoolWorker` now calls `CheckTxInputs` while before it didn't do that. So that check seemed to be redundant there. I hope I'm not wrong about that.

instagibbs commented:
That was something you had added, right?

stevenroose commented:
@instagibbs yeah yeah of course. I added it by porting back, but then realized I added the check twice since it was also in `CheckTxInputs`. elements-0.14.1 has the pegin double-pegin-spend check separate: in `CheckTxInputs` for `ConnectTip` and separate independent for `AcceptToMempoolWorker`.

instagibbs commented:
i poked the integration tests so I can see results outside of linter's complaints

instagibbs commented:
```
pegins.cpp: In function ‘bool IsValidPeginWitness(const CScriptWitness&, const COutPoint&, bool)’:
pegins.cpp:317:37: warning: ‘num_txs’ may be used uninitialized in this function [-Wmaybe-uninitialized]
         if (!IsConfirmedBitcoinBlock(block_hash, Params().GetConsensus().pegin_min_depth, num_txs)) {
                                     ^
pegins.cpp:263:9: note: ‘num_txs’ was declared here
     int num_txs;
```

instagibbs commented on src/pegins.cpp:
```diff
@@ -0,0 +1,333 @@
+    return true;
+}
+
+// Constructs unblinded output to be used in amount and scriptpubkey checks during pegin
```
suggested rewording: Constructs unblinded "bitcoin" output to be used in amount and scriptpubkey checks during pegin validation

stevenroose replied:
done

instagibbs commented on src/pegins.cpp:
```diff
@@ -0,0 +1,333 @@
+
+    //TODO(rebase) CA
+    //// Get asset type
+    //if (stack[1].size() != 32) {
```
Check the size regardless. Make sure it's size 0 for now? Or 32 exactly?

stevenroose replied:
done

instagibbs commented on src/wallet/rpcwallet.cpp:
```diff
@@ -4753,6 +4762,449 @@ UniValue walletcreatefundedpsbt(const JSONRPCRequest& request)
+    //TODO(rebase) CT/CA
+    //  v this line was in elements-0.14 instead of the line above here
+    ////SendMoney(scriptPubKey, nAmount, Params().GetConsensus().pegged_asset, subtract_fee, CPubKey(), wtxNew, true);
+    //std::string blinds;
```
you can delete these logging lines

stevenroose replied:
done

instagibbs commented on src/wallet/rpcwallet.cpp:
```diff
@@ -4753,6 +4762,449 @@ UniValue walletcreatefundedpsbt(const JSONRPCRequest& request)
+    EnsureWalletIsUnlocked(pwallet);
+
+    mapValue_t mapValue;
+    CCoinControl no_coin_control; // This is a deprecated API
```
what part is deprecated?

instagibbs commented on test/functional/feature_fedpeg.py:
```diff
@@ -0,0 +1,420 @@
+
+        sidechain.sendtomainchain(some_btc_addr, bal_2, True)
+
+        #TODO(rebase) CA bitcoin balance
```
just test for 0-balance here.

stevenroose replied:
done

instagibbs commented:
Just noting I don't think peg-outs will succeed in being "standard" unless they are under 49 bytes or so in scriptPubKey length, which is fine for us.

`MAX_OP_RETURN_RELAY = 83`, `80 bytes of data, +1 for OP_RETURN,
  +2 for the pushdata opcodes.`

This will have to be relaxed for the peg-out proof(PAK) case, of course.

instagibbs commented on src/wallet/rpcwallet.cpp:
```diff
@@ -4818,6 +5265,12 @@ static const CRPCCommand commands[] =
+    // ELEMENTS:
+    { "wallet",             "getpeginaddress",                  &getpeginaddress,               {} },
+    { "wallet",             "claimpegin",                       &claimpegin,                    {"bitcoinT", "txoutproof", "claim_script"} },
+    { "wallet",             "createrawpegin",                   &createrawpegin,                {"bitcoinT", "txoutproof", "claim_script"} },
```
"bitcoinT"

stevenroose replied:
both fixed

instagibbs commented on src/wallet/rpcwallet.cpp:
```diff
@@ -4818,6 +5265,12 @@ static const CRPCCommand commands[] =
     { "wallet",             "sethdseed",                        &sethdseed,                     {"newkeypool","seed"} },
+    // ELEMENTS:
+    { "wallet",             "getpeginaddress",                  &getpeginaddress,               {} },
+    { "wallet",             "claimpegin",                       &claimpegin,                    {"bitcoinT", "txoutproof", "claim_script"} },
```
"bitcoinT"

instagibbs commented:
Could you replace all the `push_back(Pair...` with `pushKV`? This is the upstream standard now for UniValue.

stevenroose commented:
> Could you replace all the push_back(Pair... with pushKV? This is the upstream standard now for UniValue.

Whoops, I changed most, but didn't do a grep to catch ones that I forgot. Fixed now.

instagibbs commented:
that one functional test build is failing due to that assert still. Going to review anyways while we figure out the solution.

instagibbs commented on src/chainparams.cpp:
```diff
@@ -517,12 +502,61 @@ class CCustomParams : public CRegTestParams {
+
+        base58Prefixes[PARENT_PUBKEY_ADDRESS] = std::vector<unsigned char>(1, args.GetArg("-parentpubkeyprefix", 111));
+        base58Prefixes[PARENT_SCRIPT_ADDRESS] = std::vector<unsigned char>(1, args.GetArg("-parentscriptprefix", 196));
+        parent_bech32_hrp = args.GetArg("-parent_bech32_hrp", "bc");
```
https://github.com/bitcoin/bitcoin/issues/12314#issuecomment-375226104

should be regtest default :P

instagibbs commented on src/rpc/blockchain.cpp:
```diff
@@ -2212,6 +2212,55 @@ UniValue scantxoutset(const JSONRPCRequest& request)
+            "  \"pegged_asset\" : \"xxxx\",        (string) Pegged asset type in hex\n"
+            "  \"min_peg_diff\" : \"xxxx\",        (string) The minimum difficulty parent chain header target. Peg-in headers that have less work will be rejected as an anti-Dos measure.\n"
+            "  \"parent_blockhash\" : \"xxxx\",    (string) The parent genesis blockhash as source of pegged-in funds.\n"
+            "  \"parent_chain_has_pow\": \"xxxx\", (boolean) Whether parent chain has pow or signed blocks.\n"
```
answer: it did

stevenroose replied:
Yeah I also uncommented the `parent_chain_signblock_script` thing in a later commit after rebasing on the signed blocks work :)

stevenroose commented on src/pow.h:
```diff
@@ -20,4 +20,8 @@ unsigned int CalculateNextWorkRequired(const CBlockIndex* pindexLast, int64_t nF
 
+// ELEMENTS:
+/** Check whether a parent chain block hash satisfies the proof-of-work requirement specified by nBits */
+bool CheckParentProofOfWork(uint256 hash, unsigned int nBits, const Consensus::Params&);
```
or in pegins.h to keep it with pegin validation logic?

instagibbs replied:
true

stevenroose commented on src/test/pegin_witness_tests.cpp:
```diff
@@ -0,0 +1,136 @@
+    ParseHex("000000205e3913a320cd2e3a2efa141e47419f54cb9e82320cf8dbc812fc19b9a1b2413a57f5e9fb4fa22de191454a241387f5d10cc794ee0fbf72ae2841baf3129a4eab8133025affff7f20000000000200000002f9d0be670007d38fceece999cb6144658a99c307ccc37f6d8f69129ed0f4545ff321df9790633bc33c67239c4174df8142ee616ee6a2e2788fe4820fe70e9bce0105")
+};
+
+//std::vector<unsigned char> pegin_transaction = ParseHex("020000000101f321df9790633bc33c67239c4174df8142ee616ee6a2e2788fe4820fe70e9bce0100004000ffffffff0201e48a1a02a8f799892fda58347c2d794144311d4307dbfd10f77ffe28088c60be01000000003b9ab2e0001976a914809326f7628dc976fbe63806479a1b8dfcc8c4b988ac01e48a1a02a8f799892fda58347c2d794144311d4307dbfd10f77ffe28088c60be0100000000000017200000000000000000024830 ...
```
Yeah, ugh, I manually edited that tx to move the pegin witness to another place.. I thought it might be worth it to keep the original when we change the TxWitness thing for backwards compatibility.. Even though we could just get it from the elements-0.14.1 branch if we need it..

stevenroose commented on src/coins.cpp:
```diff
@@ -236,6 +236,9 @@ bool CCoinsViewCache::HaveInputs(const CTransaction& tx) const
 {
     if (!tx.IsCoinBase()) {
         for (unsigned int i = 0; i < tx.vin.size(); i++) {
+            if (tx.vin[i].m_is_pegin) {
```
I'll change this on squash. Please first review.

instagibbs commented:
```
./pegins.h:18:69: error: ‘Consensus’ does not name a type
 bool CheckParentProofOfWork(uint256 hash, unsigned int nBits, const Consensus::Params&);
```

instagibbs commented on src/txdb.cpp:
```diff
@@ -110,11 +118,19 @@ bool CCoinsViewDB::BatchWrite(CCoinsMap &mapCoins, const uint256 &hashBlock) {
+                else
+                    batch.Write(std::make_pair(DB_PEGIN_FLAG, it->first), '1');
+            } else {
+                CoinEntry entry(&it->first.second);
```
Make a note here that non-pegin entries are entered into the DB the same way as in Core.

instagibbs commented on src/txdb.cpp:
```diff
@@ -110,11 +118,19 @@ bool CCoinsViewDB::BatchWrite(CCoinsMap &mapCoins, const uint256 &hashBlock) {
+                if (!it->second.peginSpent)
+                    batch.Erase(std::make_pair(DB_PEGIN_FLAG, it->first));
+                else
+                    batch.Write(std::make_pair(DB_PEGIN_FLAG, it->first), '1');
```
Note that the value here is a static value, since all we care about is spentness of a peg-in outpoint, and nothing else(no values, scriptpubkeys, etc).

instagibbs commented on src/txdb.cpp:
```diff
@@ -110,11 +118,19 @@ bool CCoinsViewDB::BatchWrite(CCoinsMap &mapCoins, const uint256 &hashBlock) {
-            else
-                batch.Write(entry, it->second.coin);
+            if (it->second.flags & CCoinsCacheEntry::PEGIN) {
+                if (!it->second.peginSpent)
```
let's put brackets around these multi-line conditionals.

instagibbs commented on src/txdb.cpp:
```diff
@@ -110,11 +118,19 @@ bool CCoinsViewDB::BatchWrite(CCoinsMap &mapCoins, const uint256 &hashBlock) {
-                batch.Erase(entry);
-            else
-                batch.Write(entry, it->second.coin);
+            if (it->second.flags & CCoinsCacheEntry::PEGIN) {
```
Could you "mark" this whole section as Elements code in a comment?

instagibbs commented:
utACK https://github.com/ElementsProject/elements/pull/458/commits/aee6880feeca47d27a957192d263dbad4cdc7729

that intermittant test failure is really hard to trigger and is an important exception to force the user to restart important services anyways. I'm not going to let that be a blocker for merge.

instagibbs commented:
re-utACK https://github.com/ElementsProject/elements/pull/458/commits/d358364a365a87c0d5bde5d9cae7e0e6092f4f9d
